crates/encoding/src/path.rs

```diff
@@ -2,10 +2,184 @@
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
 use bytemuck::{Pod, Zeroable};
-use peniko::kurbo::Shape;
+use peniko::{
+    kurbo::{Cap, Join, Shape, Stroke},
+    Fill,
+};
 
 use super::Monoid;
 
+/// Data structure encoding stroke or fill style.
+#[derive(Clone, Copy, Debug, Zeroable, Pod, Default, PartialEq)]
+#[repr(C)]
+pub struct Style {
+    /// Encodes the stroke and fill style parameters. This field is split into two 16-bit
+    /// parts:
+    ///
+    /// - flags: u16 - encodes fill vs stroke, even-odd vs non-zero fill mode for fills and cap
+    ///                and join style for strokes. See the FLAGS_* constants below for more
+    ///                information.
+    /// ```text
+    /// flags: |style|fill|join|start cap|end cap|reserved|
+    ///  bits:  0     1    2-3  4-5       6-7     8-15
+    /// ```
+    ///
+    /// - miter_limit: u16 - The miter limit for a stroke, encoded in binary16 (half) floating
+    ///                      point representation. This field is on meaningful for the `Join::Miter`
+    ///                      join style. It's ignored for other stroke styles and fills.
+    pub flags_and_miter_limit: u32,
+
+    /// Encodes the stroke width. This field is ignored for fills.
+    pub line_width: f32,
+}
+
+impl Style {
+    /// 0 for a fill, 1 for a stroke
+    pub const FLAGS_STYLE_BIT: u32 = 0x8000_0000;
+
+    /// 0 for non-zero, 1 for even-odd
+    pub const FLAGS_FILL_BIT: u32 = 0x4000_0000;
+
+    /// Encodings for join style:
+    ///    - 0b00 -> bevel
+    ///    - 0b01 -> miter
+    ///    - 0b10 -> round
+    pub const FLAGS_JOIN_BITS_BEVEL: u32 = 0;
+    pub const FLAGS_JOIN_BITS_MITER: u32 = 0x1000_0000;
+    pub const FLAGS_JOIN_BITS_ROUND: u32 = 0x2000_0000;
+
+    #[cfg(test)]
+    pub const FLAGS_JOIN_MASK: u32 = 0x3000_0000;
+
+    /// Encodings for cap style:
+    ///    - 0b00 -> butt
+    ///    - 0b01 -> square
+    ///    - 0b10 -> round
+    pub const FLAGS_START_CAP_BITS_BUTT: u32 = 0;
+    pub const FLAGS_START_CAP_BITS_SQUARE: u32 = 0x0400_0000;
+    pub const FLAGS_START_CAP_BITS_ROUND: u32 = 0x0800_0000;
+    pub const FLAGS_END_CAP_BITS_BUTT: u32 = 0;
+    pub const FLAGS_END_CAP_BITS_SQUARE: u32 = 0x0100_0000;
+    pub const FLAGS_END_CAP_BITS_ROUND: u32 = 0x0200_0000;
+
+    pub const FLAGS_START_CAP_MASK: u32 = 0x0C00_0000;
+    pub const FLAGS_END_CAP_MASK: u32 = 0x0300_0000;
+    pub const MITER_LIMIT_MASK: u32 = 0xFFFF;
+
+    pub fn from_fill(fill: &Fill) -> Self {
```

Comment:
I think it’d be slightly cleaner to take this parameter by value.

Reply:
Done.